<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>Adaptive Shrinkage of correlations using CorShrink</title>

<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<link rel="stylesheet" href="cosmo.css" type="text/css" />

</head>

<body>


<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <link href="https://fonts.googleapis.com/css?family=Open+Sans:100,300,400,600" rel="stylesheet" type="text/css">
    <link href="../docs/cosmo.css" rel="stylesheet">
    <link href="../docs/cosmo.min.css" rel="stylesheet">
</head>
<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
<script src="../docs/bootstrap/js/bootstrap.min.js"></script>

<body>

<div class = "nav" align = "middle">
<div class= "p1"></div>
<p>
  <a class="btn btn-large btn btn-primary" type="button" href = "index.html">CorShrink</a>
  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp; &nbsp; &nbsp;
  <a class="btn btn-large btn btn-primary" type="button" href = "workflow.html">DEMO</a>
  &nbsp &nbsp
  <a class="btn btn-large btn btn-primary" type="button" href = "workflow_project.html">WORKFLOW</a>
  &nbsp &nbsp
  <a class="btn btn-large btn btn-primary" type="button" href = "https://github.com/kkdey/CorShrink">SOFTWARE</a>
 </p>
</div>
<!-- <div class="nav">
  <a href="index.html">Home page</a> &nbsp &nbsp
  <a href="musings.html">Musings</a> &nbsp &nbsp
  <a href="town.html">My town</a> &nbsp &nbsp
  <a href="links.html">Links</a>
</div> -->

<!-- Main content -->
</body>
</html>


<h1 class="title toc-ignore">Adaptive Shrinkage of correlations using <em>CorShrink</em></h1>
<h4 class="author"><em>Kushal K Dey</em></h4>


<div id="TOC">
<ul>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#corshrink-installation">CorShrink Installation</a></li>
<li><a href="#methods">Methods</a></li>
<li><a href="#illustration">Illustration</a><ul>
<li><a href="#standard-version-corshrink">Standard version <em>CorShrink</em></a><ul>
<li><a href="#corshrinkdata"><em>CorShrinkData</em></a></li>
<li><a href="#corshrinkmatrix"><em>CorShrinkMatrix</em></a></li>
<li><a href="#corshrinkvector"><em>CorShrinkVector</em></a></li>
</ul></li>
<li><a href="#re-sampling-version-corshrink">Re-sampling version <em>CorShrink</em></a><ul>
<li><a href="#corshrinkdata---resampling"><em>CorShrinkData</em> - resampling</a></li>
<li><a href="#corshrinkmatrix---resampling"><em>CorShrinkMatrix</em> - resampling</a></li>
</ul></li>
</ul></li>
<li><a href="#extras">Extras</a></li>
<li><a href="#acknowledgements">Acknowledgements</a></li>
<li><a href="#session-info">Session Info</a></li>
<li><a href="#references">References</a></li>
</ul>
</div>

<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>Estimation of covariance or correlation matrices has widespread usage in a broad spectrum of statistical applications. The most commonly used estimator, namely the sample covariance or correlation matrix, is rank deficient and hence unstable in cases where the dimensionality of the problem (p) is greater than the number of samples (n). This problem has driven statisticians to suggest various alternative estimators in  settings. Several estimators of correlation matrix have been proposed in such settings and their theoretical properties and performance comparisons have been studied comprehensively <span class="citation">[@touloumis2015nonparametric, @ledoit2003improved, @bickel2008, @rothman2009]</span>. Some of these methods are already available as R packages - <strong>corpcor</strong> <span class="citation">[@schafer2004empirical]</span>, <strong>glasso</strong> <span class="citation">[@friedman2008sparse]</span>, <strong>PDSCE</strong> <span class="citation">[@rothman2009]</span> etc.</p>
<p>These approaches, however, are not well suited for handling large scale missingness in data. Also, some of these methods work well under some specific sets of assumptions about the underlying matrix, for e.g. - thresholding estimators assume a banded structure of the correlation matrix. In this package, ee introduce a method <em>CorShrink</em> that adapts to varying degree of missingness in observations corresponding to each pair of features. Also, <em>CorShrink</em> can be applied directly to data consisting of missing values, as well as to derived quantities like vectors and matrices of correlations between features, and allows for two formulations - an asymptotic approach and a resampling based approach. Even in examples with no missing data, CorShrink estimated correlations are visibly closer to the true correlations compared to the standard methods. <em>CorShrink</em> also can be applied to other correlation-like quantities such as partial correlations, rank correlations and cosine similarity values from word2vec models.</p>
<hr />
</div>
<div id="corshrink-installation" class="section level1">
<h1>CorShrink Installation</h1>
<p><strong>CorShrink</strong> is a companion package to <strong>ashr</strong> R package <span class="citation">[@stephens2016false]</span>. Before installing <strong>CorShrink</strong>, please make sure you have the latest version of <strong>ashr</strong>.</p>
<pre class="r"><code>devtools::install_github(&quot;stephens999/ashr&quot;)</code></pre>
<p>The other dependencies of this package include <strong>SQUAREM</strong>, <strong>reshape2</strong> and <strong>Matrix</strong>. Next we install <strong>CorShrink</strong>.</p>
<pre class="r"><code>install.packages(&quot;CorShrink&quot;)</code></pre>
<p>The development version can be installed from Github as well.</p>
<pre class="r"><code>library(devtools)
install_github(&quot;kkdey/CorShrink&quot;)</code></pre>
<p>Then load the package with:</p>
<pre class="r"><code>library(CorShrink)</code></pre>
<hr />
</div>
<div id="methods" class="section level1">
<h1>Methods</h1>
<p>The main steps in <strong>CorShrink</strong> are as follows</p>
<ul>
<li>Convert the correlation between variables <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span>, <span class="math inline">\(R_{ij}\)</span> into Fisher z-score <span class="math inline">\(Z_{ij}\)</span> by the following transformation.</li>
</ul>
<p><span class="math display">\[  Z_{ij} = 0.5 \log \left (\frac{1 + R_{ij}}{1 - R_{ij}} \right )  \]</span></p>
<ul>
<li><p>Estimate from data, the standard errors (<span class="math inline">\(s_{ij}\)</span>) of these Z-scores <span class="math inline">\(Z_{ij}\)</span>. This can be done in two ways in <strong>CorShrink</strong>. One approach uses an asymptotic normal approximation, where the standard errors are <span class="math inline">\(s_{ij} = \frac{1}{n_{ij} - 3}\)</span>, with <span class="math inline">\(n_{ij}\)</span> being the number of complete observations between pair <span class="math inline">\((i,j)\)</span> that generates the correlation <span class="math inline">\(R_{ij}\)</span>. The other approach performs a re-sampling of the observations for the <span class="math inline">\((i,j)\)</span> pair and obtains a Bootstrap estimate of the standard errors from the re-sampled <span class="math inline">\(Z_{ij}\)</span>.</p></li>
<li><p>Apply adaptive shrinkage (<em>ash</em> due <span class="citation">[@stephens2016false]</span>) on the pairs <span class="math inline">\((Z_{ij}, s_{ij})\)</span> either across all <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> pairs (matrix format) or along all <span class="math inline">\(i\)</span> for one <span class="math inline">\(j\)</span>, or along all <span class="math inline">\(j\)</span> for one <span class="math inline">\(i\)</span> (vector formats).</p></li>
</ul>
<p><span class="math display">\[  Z^{\star}_{ij}  : = ash \; (Z_{ij}, s_{ij})   \]</span></p>
<p>The matrix format shrinkage is performed by the <em>CorShrinkMatrix</em> function while the vector format shrinkage is performed by the <em>CorShrinkVector</em> function.</p>
<ul>
<li>Reverse transform the posterior mean of the shrunk Fisher z-scores from the previous step.</li>
</ul>
<p><span class="math display">\[  R^{\star}_{ij} = \frac{exp \; (2 Z^{\star}_{ij}) - 1}{exp \; (2 Z^{\star}_{ij}) + 1} \]</span></p>
<ul>
<li>If the user is attempting to estimate a correlation matrix ((R_{ij})) for a number of variables <span class="math inline">\((i,j)\)</span>, then a requisite condition is positive definiteness of the estimate. However, the <span class="math inline">\(((R^{\star}_{ij}))\)</span> matrix obtained above may not be positive definite. We work around this problem by choosing a positive definite matrix <span class="math inline">\(((R^{{\star}{\star}}_{ij}))\)</span> nearest to the above estimate (see <span class="citation">[@higham2002computing]</span>).</li>
</ul>
<hr />
</div>
<div id="illustration" class="section level1">
<h1>Illustration</h1>
<p>We load an example data matrix - the person (544) by tissue samples (53) gene expression data for the gene <em>ENSG00000166819</em> collected from the <a href="https://www.gtexportal.org/home/">Genotype Tissue Expression (GTEx) Project</a> .</p>
<pre class="r"><code>data(&quot;sample_by_feature_data&quot;)</code></pre>
<p>Just by checking the first few rows and columns, we see that the data contains many missing values. The data is</p>
<pre class="r"><code>sample_by_feature_data[1:5,1:5]</code></pre>
<pre><code>##            Adipose - Subcutaneous Adipose - Visceral (Omentum)
## GTEX-111CU              10.472332                     10.84006
## GTEX-111FC               7.335392                           NA
## GTEX-111VG               9.118889                           NA
## GTEX-111YS              10.806459                     11.26113
## GTEX-1122O              11.040446                     11.71497
##            Adrenal Gland Artery - Aorta Artery - Coronary
## GTEX-111CU      2.721234             NA                NA
## GTEX-111FC            NA             NA                NA
## GTEX-111VG            NA             NA                NA
## GTEX-111YS      3.454823       1.162059                NA
## GTEX-1122O      1.522667       1.674467          4.188002</code></pre>
<div id="standard-version-corshrink" class="section level2">
<h2>Standard version <em>CorShrink</em></h2>
<div id="corshrinkdata" class="section level3">
<h3><em>CorShrinkData</em></h3>
<p>We estimate the adaptively shrunk correlation matrix for this data using <strong>CorShrink</strong>.</p>
<pre class="r"><code>out &lt;- CorShrinkData(sample_by_feature_data, sd_boot = FALSE, image = &quot;both&quot;,
                     image.control = list(tl.cex = 0.8))</code></pre>
<p><img src="corshrink_files/figure-html/corshrink_data-1.png" width="900" style="display: block; margin: auto auto auto 0;" /></p>
<p>The function outputs a list with two elements which are two versions of CorShrink estimated matrices - <code>cor</code> and <code>cor_before_PD</code>.<code>cor</code> is the nearest positive definite approximation (<span class="math inline">\(R^{{\star}{\star}}\)</span>) to <code>cor_before_PD</code> version (<span class="math inline">\(R^{{\star}}\)</span>) as described in the methods above. When <code>image = &quot;both&quot;</code>, the function plots the images for both these versions.</p>
<p>To see whether the method works well, check if the these two versions are close to each other.</p>
<pre class="r"><code>out$cor_before_PD[1:5,1:5]</code></pre>
<pre><code>##                              Adipose - Subcutaneous
## Adipose - Subcutaneous                   1.00000000
## Adipose - Visceral (Omentum)             0.24167719
## Adrenal Gland                           -0.04364913
## Artery - Aorta                           0.01307030
## Artery - Coronary                        0.21743134
##                              Adipose - Visceral (Omentum) Adrenal Gland
## Adipose - Subcutaneous                        0.241677186  -0.043649133
## Adipose - Visceral (Omentum)                  1.000000000   0.002099297
## Adrenal Gland                                 0.002099297   1.000000000
## Artery - Aorta                                0.004398573  -0.001082796
## Artery - Coronary                             0.012282042   0.038295685
##                              Artery - Aorta Artery - Coronary
## Adipose - Subcutaneous          0.013070305        0.21743134
## Adipose - Visceral (Omentum)    0.004398573        0.01228204
## Adrenal Gland                  -0.001082796        0.03829568
## Artery - Aorta                  1.000000000        0.03868019
## Artery - Coronary               0.038680191        1.00000000</code></pre>
<pre class="r"><code>out$cor[1:5, 1:5]</code></pre>
<pre><code>##                              Adipose - Subcutaneous
## Adipose - Subcutaneous                   1.00000000
## Adipose - Visceral (Omentum)             0.24112346
## Adrenal Gland                           -0.04174961
## Artery - Aorta                           0.01346126
## Artery - Coronary                        0.21567603
##                              Adipose - Visceral (Omentum) Adrenal Gland
## Adipose - Subcutaneous                        0.239767447   -0.04149049
## Adipose - Visceral (Omentum)                  1.000000000    0.00169961
## Adrenal Gland                                 0.001700607    1.00000000
## Artery - Aorta                                0.003230271   -0.00079904
## Artery - Coronary                             0.013433817    0.03748903
##                              Artery - Aorta Artery - Coronary
## Adipose - Subcutaneous         0.0133903272        0.21453361
## Adipose - Visceral (Omentum)   0.0032314217        0.01343823
## Adrenal Gland                 -0.0007997934        0.03752334
## Artery - Aorta                 1.0000000000        0.03828987
## Artery - Coronary              0.0382909260        1.00000000</code></pre>
</div>
<div id="corshrinkmatrix" class="section level3">
<h3><em>CorShrinkMatrix</em></h3>
<p><strong>CorShrink</strong> takes as input not just the samples by features data matrix but also a matrix of pairwise correlations with a matrix of number of samples for each pair contributing to the correlation.</p>
<pre class="r"><code>data(&quot;pairwise_corr_matrix&quot;)
data(&quot;common_samples&quot;)
out &lt;- CorShrinkMatrix(pairwise_corr_matrix, common_samples, image = &quot;both&quot;,
                       image.control = list(tl.cex = 0.8))</code></pre>
<p><img src="corshrink_files/figure-html/corshrinkmat-1.png" width="900" style="display: block; margin: auto auto auto 0;" /></p>
</div>
<div id="corshrinkvector" class="section level3">
<h3><em>CorShrinkVector</em></h3>
<p><strong>CorShrink</strong> can be applied to vectors of correlations as well.</p>
<pre class="r"><code>cor_vec &lt;- c(-0.56, -0.4, 0.02, 0.2, 0.9, 0.8, 0.3, 0.1, 0.4)
nsamp_vec &lt;- c(10, 20, 30, 4, 50, 60, 20, 10, 3)
out &lt;- CorShrinkVector(corvec = cor_vec, nsamp_vec = nsamp_vec)
out</code></pre>
<pre><code>## [1] -0.1008065570 -0.0593108888  0.0006072387  0.0127953383  0.8944339076
## [6]  0.7935595323  0.0236537504  0.0051443096  0.0250218532</code></pre>
<p>Note that the correlations computed from adequate amount of data as for the 5th and 6th entries above, the amount of shrinkage is minimal, while it is substantial for the 4th and 9th entries which correspond to small number of samples.</p>
</div>
</div>
<div id="re-sampling-version-corshrink" class="section level2">
<h2>Re-sampling version <em>CorShrink</em></h2>
<p>We have so far looked at <em>CorShrinkData</em>, <em>CorShrinkMatrix</em> and <em>CorShrinkVector</em>, three functions that provide adaptive shrinkage of correlations at the level of the data matrix, matrix of correlations and vector of correlations respectively. In the above examples, we have used the asymptotic version of our algorithm (see Methods). Next we show example usage of a resampling based version of <em>CorShrink</em>.</p>
<div id="corshrinkdata---resampling" class="section level3">
<h3><em>CorShrinkData</em> - resampling</h3>
<pre class="r"><code>out &lt;- CorShrinkData(sample_by_feature_data, sd_boot = TRUE, image = &quot;both&quot;,
                     image.control = list(tl.cex = 0.8))</code></pre>
<pre><code>## Finished Bootstrap : 1
## Finished Bootstrap : 2
## Finished Bootstrap : 3
## Finished Bootstrap : 4
## Finished Bootstrap : 5
## Finished Bootstrap : 6
## Finished Bootstrap : 7
## Finished Bootstrap : 8
## Finished Bootstrap : 9
## Finished Bootstrap : 10
## Finished Bootstrap : 11
## Finished Bootstrap : 12
## Finished Bootstrap : 13
## Finished Bootstrap : 14
## Finished Bootstrap : 15
## Finished Bootstrap : 16
## Finished Bootstrap : 17
## Finished Bootstrap : 18
## Finished Bootstrap : 19
## Finished Bootstrap : 20
## Finished Bootstrap : 21
## Finished Bootstrap : 22
## Finished Bootstrap : 23
## Finished Bootstrap : 24
## Finished Bootstrap : 25
## Finished Bootstrap : 26
## Finished Bootstrap : 27
## Finished Bootstrap : 28
## Finished Bootstrap : 29
## Finished Bootstrap : 30
## Finished Bootstrap : 31
## Finished Bootstrap : 32
## Finished Bootstrap : 33
## Finished Bootstrap : 34
## Finished Bootstrap : 35
## Finished Bootstrap : 36
## Finished Bootstrap : 37
## Finished Bootstrap : 38
## Finished Bootstrap : 39
## Finished Bootstrap : 40
## Finished Bootstrap : 41
## Finished Bootstrap : 42
## Finished Bootstrap : 43
## Finished Bootstrap : 44
## Finished Bootstrap : 45
## Finished Bootstrap : 46
## Finished Bootstrap : 47
## Finished Bootstrap : 48
## Finished Bootstrap : 49
## Finished Bootstrap : 50</code></pre>
<p><img src="corshrink_files/figure-html/corshrink_data_boot-1.png" width="900" style="display: block; margin: auto auto auto 0;" /></p>
<p>The algorithm works by first computing a Bootstrap estimate of the standard error of the Fisher z-scores for each pair and then using this estimate together with the correlations to shrink the latter.</p>
</div>
<div id="corshrinkmatrix---resampling" class="section level3">
<h3><em>CorShrinkMatrix</em> - resampling</h3>
<p>The breakdown can be formulated at the level of a correlation matrix as follows.</p>
<pre class="r"><code>zscoreSDmat &lt;- bootcorSE_calc(sample_by_feature_data, verbose = FALSE)
out &lt;- CorShrinkMatrix(pairwise_corr_matrix, zscore_sd = zscoreSDmat, image = &quot;both&quot;,
                       image.control = list(tl.cex = 0.8))</code></pre>
<p><img src="corshrink_files/figure-html/corshrink_mat_boot-1.png" width="900" style="display: block; margin: auto auto auto 0;" /></p>
<hr />
</div>
</div>
</div>
<div id="extras" class="section level1">
<h1>Extras</h1>
<p>So far, in all our examples, we assumed that the estimated correlations between any pair of variables is shrunk towards 0. But <em>CorShrink</em> allows the user to choose a non-zero shrinkage target, estimated from the data, using the <code>mode</code> option in <code>ash.control</code> input.</p>
<p>One can choose a fixed non-zero target in <code>mode</code> as well.</p>
<pre class="r"><code>par(mfrow=c(1,2))
out1 &lt;- CorShrinkData(sample_by_feature_data, sd_boot = FALSE, image = &quot;corshrink&quot;,             image.control = list(title = &quot;CorShrink (target = 0)&quot;, tl.cex = 0.8))
out2 &lt;- CorShrinkData(sample_by_feature_data, sd_boot = FALSE, image = &quot;corshrink&quot;,                      ash.control = list(mode = &quot;estimate&quot;),
    image.control = list(title = &quot;CorShrink (target = estimated)&quot;, tl.cex = 0.8))</code></pre>
<p><img src="corshrink_files/figure-html/extras_1-1.png" width="900" style="display: block; margin: auto auto auto 0;" /></p>
<p>The <code>image = &quot;output&quot;</code> option just outputs the image for the shrunk matrix without plotting it.</p>
<p>In general, <em>CorShrink</em> assumes a normal prior for the population Fisher z-scores. But under specific settings, a non-symmetric distribution , such as uniform or half-uniform could be a better fit. This can be achieved using the <code>mixcompdist</code> in <code>ash.control</code>.</p>
<pre class="r"><code>par(mfrow=c(2,2))
out1 &lt;- CorShrinkData(sample_by_feature_data,sd_boot = FALSE, image =&quot;corshrink&quot;,
                     ash.control = list(mixcompdist = &quot;normal&quot;),
                     image.control = list(tl.cex = 0.6))
out2 &lt;- CorShrinkData(sample_by_feature_data,sd_boot = FALSE, image =&quot;corshrink&quot;,
                     ash.control = list(mixcompdist = &quot;uniform&quot;),
                     image.control = list(tl.cex = 0.6))
out3 &lt;- CorShrinkData(sample_by_feature_data,sd_boot = FALSE, image =&quot;corshrink&quot;,
                     ash.control = list(mixcompdist = &quot;halfuniform&quot;),
                     image.control = list(tl.cex = 0.6))
out4 &lt;- CorShrinkData(sample_by_feature_data,sd_boot = FALSE, image =&quot;corshrink&quot;,
                     ash.control = list(mixcompdist = &quot;+uniform&quot;),
                     image.control = list(tl.cex = 0.6))</code></pre>
<p><img src="corshrink_files/figure-html/extras_2-1.png" width="900" style="display: block; margin: auto auto auto 0;" /></p>
<hr />
</div>
<div id="acknowledgements" class="section level1">
<h1>Acknowledgements</h1>
<p>We would like to thank the GTEx Consortium, John Blischak, Sarah Urbut, Chiaowen Joyce Hsiao, Peter Carbonetto and all members of the Stephens Lab.</p>
<hr />
</div>
<div id="session-info" class="section level1">
<h1>Session Info</h1>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 3.5.0 (2018-04-23)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS Sierra 10.12.6
##
## Matrix products: default
## BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib
##
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
##
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base
##
## other attached packages:
## [1] CorShrink_0.1-3 knitr_1.20
##
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.17      magrittr_1.5      MASS_7.3-49
##  [4] doParallel_1.0.11 pscl_1.5.2        SQUAREM_2017.10-1
##  [7] lattice_0.20-35   foreach_1.4.4     plyr_1.8.4
## [10] ashr_2.2-7        stringr_1.3.1     tools_3.5.0
## [13] corrplot_0.84     parallel_3.5.0    grid_3.5.0
## [16] gtable_0.2.0      corpcor_1.6.9     htmltools_0.3.6
## [19] iterators_1.0.9   yaml_2.1.19       rprojroot_1.3-2
## [22] digest_0.6.15     Matrix_1.2-14     gridExtra_2.3
## [25] reshape2_1.4.3    codetools_0.2-15  evaluate_0.10.1
## [28] rmarkdown_1.9     stringi_1.2.2     compiler_3.5.0
## [31] backports_1.1.2   truncnorm_1.0-8</code></pre>
<hr />
</div>
<div id="references" class="section level1">
<h1>References</h1>
</div>

<hr>
<p>
    This <a href="http://rmarkdown.rstudio.com">R Markdown</a> site was created with <a href="https://github.com/jdblischak/workflowr">workflowr</a>
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
